cavern/src/main/java/org/opencadc/cavern/nodes/PosixIdentityManager.java

```diff
@@ -176,15 +199,30 @@ public Object toOwner(Subject subject) {
     @Override
     public String toDisplayString(Subject subject) {
         // delegate to configured IM
-        IdentityManager im = AuthenticationUtil.getIdentityManager();
-        return im.toDisplayString(subject);
+        return this.wrappedIdentityManager.toDisplayString(subject);
     }
 
     @Override
     public Subject validate(Subject subject) throws NotAuthenticatedException {
-        // delegate to configured IM
-        IdentityManager im = AuthenticationUtil.getIdentityManager();
-        return im.validate(subject);
+        if (subject != null) {
+            final Set<AuthorizationTokenPrincipal> tokenPrincipals = subject.getPrincipals(AuthorizationTokenPrincipal.class);
+            for (final AuthorizationTokenPrincipal tokenPrincipal : tokenPrincipals) {
+                final String tokenValue = tokenPrincipal.getHeaderValue();
+                final String[] parts = tokenValue.split(" ", 2);
+                if (parts.length == 2) {
+                    final String challengeType = parts[0];
+                    if (CavernConfig.ALLOCATION_API_KEY_HEADER_CHALLENGE_TYPE.equalsIgnoreCase(challengeType)) {
+                        final String apiKeyValue = parts[1].trim();
+
+                        subject.getPrincipals().remove(tokenPrincipal);
+                        subject.getPublicCredentials().add(new AuthorizationToken(challengeType, apiKeyValue, Collections.emptyList()));
+                    }
+                }
+            }
+        }
+
+        // delegate to the configured IdentityManager, assuming it can handle a null Subject
```

Comment:
I would have done the call to `wrappedIdentityManager.validate(subject)` first because most of the time that would leave the subject with 0 ATPs remaining and you wouldn't have to scan through the others just to skip them.

So I think just move this up to the top, then do the loop looking for api key ATPs.